components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TabListReducer.kt

```diff
@@ -134,13 +135,23 @@ internal object TabListReducer {
 
             is TabListAction.RestoreAction -> {
                 // Verify that none of the tabs to restore already exist
-                action.tabs.forEach { requireUniqueTab(state, it) }
+                val restoredTabs = action.tabs.toTabSessionStates()
+                restoredTabs.forEach { requireUniqueTab(state, it) }
+
+                // We are adding the restored tabs at their original indices. If the removal index
+                // for some reason is null, the tab will be restored to the end of the tab list. The
+                // removal index will be reset to null when added to the combined list.
```

Comment:
The tabs without indices are currently refactored to be placed at the end of the tab list instead of the beginning. The original comment only made sense when tabs were _exclusively_ being placed at the beginning when restored.